### ISSUE TITLE
Datepicker Calendar UI sometimes is missing days of the month.

### ISSUE THREAD

**Issue by morgluc** [mui/mui-x#9914]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Steps to reproduce 🕹

Link to live example:
https://mui.com/x/react-date-pickers/timezone/

Steps:
1. Load the page
2. Scroll to the date time picker
3. Click the left arrow UI button to scroll through months until you see February
4. Observe that February has 5 days instead of 28.
![image](https://github.com/mui/mui-x/assets/63873395/62bb29f4-7795-4334-ae85-5d92e078719f)

### Current behavior 😯

The calendar UI shows only 5 days.

### Expected behavior 🤔

The calendar UI shows 28 days.

### Context 🔦

I'm trying to select a date in a past month using the calendar UI, but the days I wish to select (January 1st, 2023) doesn't exist in the calendar UI.

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```

Using Firefox 115.0.3 64 bit

```
PS C:\Users\morgluc\source\repos\GF.RouteManagement.V3.UI> npx @mui/envinfo

  System:
    OS: Windows 10 10.0.19045
  Binaries:
    Node: 16.14.2 - C:\Program Files\nodejs\node.EXE
    Yarn: Not Found
    npm: 8.5.0 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Chrome: Not Found
    Edge: Spartan (44.19041.1266.0), Chromium (115.0.1901.188)
  npmPackages:
    @emotion/react:  11.10.8
    @emotion/styled:  11.10.8
    @mui/base:  5.0.0-alpha.128
    @mui/core-downloads-tracker:  5.12.3
    @mui/icons-material:  5.11.16
    @mui/material: ^5.12.2 => 5.12.3
    @mui/private-theming:  5.12.3
    @mui/styled-engine:  5.12.3
    @mui/system:  5.12.3
    @mui/types:  7.2.4
    @mui/utils:  5.14.1
    @mui/x-date-pickers: ^6.10.1 => 6.10.1
    @types/react: ^17.0.58 => 17.0.58
    react: ^17.0.2 => 17.0.2
    react-dom: ^17.0.2 => 17.0.2
    styled-components: ^6.0.0-rc.3 => 6.0.0-rc.3
    typescript: ^4.9.5 => 4.9.5

PS C:\Users\morgluc\source\repos\GF.RouteManagement.V3.UI>
```
</details>

**Comment by LukasTy** [comment]
Thank you for opening this issue! 🙏 
@flaviendelangle hasn't this already been solved yesterday with https://github.com/mui/mui-x/pull/9884?

**Comment by flaviendelangle** [comment]
I can reproduce on the latest released version and it has indeed been fixed by #9884 

Thanks for the feedback!
You should have the fix today in the new release :ok_hand:
